tests/kernel/mutex/mutex_api/src/test_mutex_apis.c

```diff
@@ -387,6 +387,52 @@ ZTEST_USER(mutex_api_1cpu, test_mutex_priority_inheritance)
 	k_msleep(TIMEOUT+1000);
 }
 
+static void tThread_mutex_lock_should_fail(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	struct k_mutex *mutex = (struct k_mutex *)p1;
+
+	zassert_equal(-EAGAIN, k_mutex_lock(mutex, K_MSEC(TIMEOUT)), NULL);
+}
+
+/**
+ * @brief Test fix for subtle race during priority inversion
+ *
+ * - A low priority thread (Tlow) locks mutex A.
+ * - A high priority thread (Thigh) blocks on mutex A, boosting the priority
+ *   of Tlow.
+ * - Thigh times out waiting for mutex A.
+ * - Before Thigh has a chance to execute, Tlow unlocks mutex A (which now
+ *   has no owner) and drops its own priority.
+ * - Thigh now gets a chance to execute and finds that it timed out, and
+ *   then enters the block of code to lower the priority of the thread that
+ *   owns mutex A (now nobody).
+ * - Thigh tries to the dereference the owner of mutex A (which is nobody,
+ *   and thus it is NULL). This leads to an exception.
+ *
+ * @ingroup kernel_mutex_tests
+ *
+ * @see k_mutex_lock()
+ */
+ZTEST(mutex_api_1cpu, test_mutex_timeout_race_during_priority_inversion)
+{
+	k_mutex_init(&mutex);
+
+	/* align to tick boundary */
+	k_sleep(K_TICKS(1));
+	k_thread_create(&tdata, tstack, K_THREAD_STACK_SIZEOF(tstack),
+			tThread_mutex_lock_should_fail, &mutex, NULL, NULL, 8,
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_mutex_lock(&mutex, K_FOREVER);
```

Comment:
Minor item: I would prefer to see that *k_mutex_lock()* is moved to before *k_thread_create()* is called. This makes it blatantly obvious to a reader that the mutex is locked before the helper thread executes. Otherwise anyone reading the code in the future has to expend additional energy/time verifying priorities, scheduling, and number of cpus involved to ensure that  it behaves as described.

Reply:
Back up a moment ... isn't the priority of the running thread here a higher priority level than the created thread? If true, then the current thread gets the thread and when the created thread tries to take the semaphore, it would not lead to a case of priority inversion.

Thus, we should be doing ...

1. lock mutex
2. create helper thread at higher priority to current thread.